esp-hal/src/tsens.rs

```diff
@@ -0,0 +1,117 @@
+//! # Temperature Sensor (tsens)
+//!
+//! ## Overview
+//!
+//! The Temperature Sensor peripheral is used to measure the internal
+//! temperature inside the chip. The voltage is internally converted via an ADC
+//! into a digital value, and has a measuring range of –40 °C to 125 °C.
+//! The temperature value depends on factors like microcontroller clock
+//! frequency or I/O load. Generally, the chip’s internal temperature is higher
+//! than the operating ambient temperature.
+//!
+//! ## Configuration
+//!
+//! Currently, it does not support any configuration, but some future
+//! configuration options are listed below in the Implementation State list of
+//! pending items
+//!
+//! ## Examples
+//!
+//! The following example will measure the internal chip temperature every
+//! second, and print it
+//!
+//! ```rust, no_run
+#![doc = crate::before_snippet!()]
+//! # use esp_hal::tsens::{TemperatureSensor, Config};
+//! # use esp_hal::delay::Delay;
+//!
+//! let tsens = TemperatureSensor::new(peripherals.TSENS);
+//! let delay = Delay::new();
+//!
+//! loop {
+//!   let temp = tsens.get_celsius();
+//!   println!("Temperature: {:.2}°C", temp);
+//!   delay.delay_millis(1_000);
+//! }
+//! # }
+//! ```
+//! 
+//! ## Implementation State
+//!
+//! - Source clock selection is not supported
+//! - Temperature calibration range is not supported
+//! - Interrupts are not supported
+
+use crate::{
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals::TSENS,
+    system::{GenericPeripheralGuard, PeripheralClockControl},
+};
+
+/// Temperature sensor driver
+// #[derive(Clone, Copy)]
+pub struct TemperatureSensor<'d> {
+    _peripheral: PeripheralRef<'d, TSENS>,
+    _guard: GenericPeripheralGuard<{ crate::system::Peripheral::Tsens as u8 }>,
+}
+
+impl<'d> TemperatureSensor<'d> {
+    /// Create a new temperature sensor instance
+    pub fn new(peripheral: impl Peripheral<P = TSENS> + 'd) -> Self {
+        crate::into_ref!(peripheral);
+
+        let guard = GenericPeripheralGuard::new();
+
+        // We need to enable ApbSarAdc clock before trying to write on the tsens_ctrl
+        // register
+        PeripheralClockControl::enable(crate::system::Peripheral::ApbSarAdc);
```

Comment:
What is the best way to handle this requirement of enabling `Peripheral::AbpSarAdc` before modifying the power up register for `tsens_ctrl`?

Reply:
I think we should also guard this with a `GenericPeripheralGuard`, as iirc the ADC uses this, and we probably don't want to reset & enable if it's already in use.